libhdf5/hdf5open.c

```diff
@@ -756,12 +756,13 @@ nc4_open_file(const char *path, int mode, void* parameters, int ncid)
 #endif /* !USE_PARALLEL4 */
 
     /* Need this access plist to control how HDF5 handles open objects
-     * on file close. (Setting H5F_CLOSE_SEMI will cause H5Fclose to
-     * fail if there are any open objects in the file). */
+     * on file close. (Setting H5F_CLOSE_WEAK will cause H5Fclose not to
+     * fail if there are any open objects in the file. This may happen when virtual
+     * datasets are opened). */
     if ((fapl_id = H5Pcreate(H5P_FILE_ACCESS)) < 0)
         BAIL(NC_EHDFERR);
 
-    if (H5Pset_fclose_degree(fapl_id, H5F_CLOSE_SEMI) < 0)
+    if (H5Pset_fclose_degree(fapl_id, H5F_CLOSE_WEAK) < 0)
```

Comment:
This is a somewhat drastic step. Can we discuss?

Can you tell me more about why a virtual file might have objects left open?

Reply:
I find the step also a bit too drastic. I don't really know about the thought process in HDF5 is regarding the implementation of virtual datasets. But if you have a look at the links above, [this line](https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5/browse/src/H5Dvirtual.c?at=refs%2Ftags%2Fhdf5-1_12_0#941) opens a new reference to the linked file and almost [immediately after that](https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5/browse/src/H5Dvirtual.c?at=refs%2Ftags%2Fhdf5-1_12_0#966) the file is closed again, but the previous reference is not released. This leads to an error if fclose_degree is SEMI. Also as the corresponding function is called [`H5D__virtual_open_source_dset`](https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5/browse/src/H5Dvirtual.c?at=refs%2Ftags%2Fhdf5-1_12_0#893), I assumed that at the end of the function something should be left in an open state... The above change was a way to get rid of the error without having to dig further into HDF5 and as h5netcdf could handle this, I thought there should be a way to handle it without changing HDF5.

But yes, I'd be happy to find a different solution.

Reply:
Did you find a different solution for this?

Reply:
I've sent a request to HDFgroup to clarify if the need for `H5F_CLOSE_WEAK` is intended behaviour or if I am missing something.

Reply:
I just wanted to briefly get back on this. It turns out that this behaviour is indeed a bug in HDF5. I just got the confirmation that there's a fix almost ready which should be released soon. This fix however effectively will force source datasets to be opened as `WEAK` internally.

This raises the question of how to proceed in this PR?
* Should there be a backwards-compatible solution which runs with current HDF5 by setting file opening `WEAK` outside of HDF5?
* Should there be a requirement to use HDF5 >= the upcoming release?